### PR TITLE
include/openssl/macros.h: Rework OPENSSL_FUNC for div C standards

### LINKED ISSUE #9911: OPENSSL_FUNC goes wrong with GNU C and -pedantic
The recent c659882c9892788085fcdd4e8c47f98c01edf9ad removed the use of  `PEDANTIC` in `include/openssl/macros.h`.
This causes issues when using configuration option `-std=gnu90` and compiling with GCC:
```
include/openssl/macros.h:145:26: error: ISO C does not support '__FUNCTION__' predefined identifier [-Werror=pedantic]

```

To fix this, I would have suggested replacing
```
#  elif defined(_MSC_VER) || (defined(__GNUC__) && __GNUC__ >= 2)
#   define OPENSSL_FUNC __FUNCTION__
```
by
```
#  elif defined(_MSC_VER)
#   define OPENSSL_FUNC __FUNCTION__
#  elif defined(__GNUC__) && __GNUC__ >= 2
#   if defined PEDANTIC
#    define OPENSSL_FUNC "(unknown function due to -pedantic)"
#   else
#    define OPENSSL_FUNC __FUNCTION__
#   endif

```
but apparently `PEDANTIC` should not be used in this header file.

BTW, when I modify `macros.h`, `make` does not re-build anything, so it looks like the dependencies on this file are not properly reflected in the `Makefile`.

### PR DESCRIPTION
**Pull request by levitte**
OPENSSL_FUNC was defined as an alias for __FUNCTION__ with new enough
GNU C, regardless of the language standard used.  We change this
slightly, so this won't happen unless __STDC_VERSION is defined.

Fixes #9911

### CODE REVIEW

**Comment by levitte** [comment]
@DDvO, please have a look

**Comment by DDvO** [comment]
As I wrote in #9911, also this version won't work when configuring with `-std=gnu90` (or `-std=c90`) and `-pedantic`.

**Comment by DDvO** [comment]
I had not tried before giving my above answer, but I've just tried, and it confirmed what I was meanwhile expecting:

- you are right in the sense that the build now succeeds
- yet this renders `OPENSSL_FUNC useless` (giving just `"(unknown function)"`) with anything before C94. Whereas without `-pedantic` it could result in a meaningful value, and this is why I consider the solution provided here not very satisfactory.

**Comment by DDvO** [comment]
Again my question: where to find an official statement which C standards are to be supported by OpenSSL code?

I meanwhile found that C90 builds break at various places, while GUN90 works fine (apart from the issue under discussion here).

**Comment by levitte** [comment]
> yet this renders `OPENSSL_FUNC useless` (giving just `"(unknown function)"`) with anything before C94. Whereas without `-pedantic` it could result in a meaningful value, and this is why I consider the solution provided here not very satisfactory.

Yes, this is a consequence, it's perfectly true.  The reason for this is that *`-pedantic` is not detectable* (and quite frankly, it shouldn't be...).  So we have to turn and look at the language standard, there's really no way around that.  Do note, though, that if you think you know better, it's still perfectly possible to define `OPENSSL_FUNC` differently:

- in OpenSSL configuration: just add `-DOPENSSL_FUNC=whateveryoulike`
- in your application: `#define OPENSSL_FUNC whateveryoulike` before including any OpenSSL header.

> Why not make `PEDANTIC` and official(ly usable) OpenSSL macro, e.g., as `OPENSSL_PEDANTIC`?

Because it basically means that any time *any application is to be compiled with `-pedantic`*, whoever does that must remember to define `OPENSSL_PEDANTIC`.  Experience shows that people generally do not remember these kinds of things, which means we'll have to deal with the occasional (or deluge of...) new issue complaining about how OpenSSL is not usable with `-pedantic`.  That would be an unnecessary waste of our resources.

Like I wrote above, the other option is to have those who want define their own `OPENSSL_FUNC`.

> Again my question: where to find an official statement which C standards are to be supported by OpenSSL code?

We require it to be possible to compile OpenSSL with C89/C90.  This means that we require our code not to use features from newer standard, *or in cases where we do*, to make it in such a way that there's a C90 alternative (see the `ossl_inline` macro).  The same goes for standard library headers and functions we use.

This question comes up a little now and then, it should probably be part of our FAQ...

**Comment by DDvO** [comment]
Thanks for your swift and to-the-point reply.

> > yet this renders `OPENSSL_FUNC useless` (giving just `"(unknown function)"`) with anything before C94. Whereas without `-pedantic` (which may be rarely used in practice) it could result in a meaningful value, and this is why I consider the solution provided here not very satisfactory.
> 
> Yes, this is a consequence, it's perfectly true. The reason for this is that _`-pedantic` is not detectable_ (and quite frankly, it shouldn't be...). So we have to turn and look at the language standard, there's really no way around that. 

Well, I find it at odd (by GCC at least) that the C version and type of the compiler used can be detected during compilation while `-pedantic` apparently cannot, though all of these have an effect on the available features.

> Do note, though, that if you think you know better, it's still perfectly possible to define `OPENSSL_FUNC` differently:
> 
>     * in OpenSSL configuration: just add `-DOPENSSL_FUNC=whateveryoulike`
> 
>     * in your application: `#define OPENSSL_FUNC whateveryoulike` before including any OpenSSL header.

Good hint. Yet I fear that most users experiencing `"(unknown function)"` with GNU/C90 would not be aware of that. Therefore, I suggest adding this as comment just after the line giving `"(unknown function)'`' in `macros.h`, e.g.:
```
/*- 
   with GCC and -std=c90 or -std=gnu90, as long as -pedantic is not used,
   one may resort to using
   #define OPENSSL_FUNC __FUNCTION__
   before including any OpenSSL headers
*/
```

> > Why not make `PEDANTIC` and official(ly usable) OpenSSL macro, e.g., as `OPENSSL_PEDANTIC`?
> 
> Because it basically means that any time _any application is to be compiled with `-pedantic`_, whoever does that must remember to define `OPENSSL_PEDANTIC`. Experience shows that people generally do not remember these kinds of things, which means we'll have to deal with the occasional (or deluge of...) new issue complaining about how OpenSSL is not usable with `-pedantic`. That would be an unnecessary waste of our resources.

I see. Though at least when using the OpenSSL configuration tools, the consistency of `(OPENSSL_)PEDANTIC` with `-pedantic` would be taken care of automatically.

**Comment by DDvO** [comment]
> > Again my question: where to find an official statement which C standards are to be supported by OpenSSL code?
> 
> We require it to be possible to compile OpenSSL with C89/C90. 

Ah, yes, I meanwhile found that on https://www.openssl.org/policies/codingstyle.html:
```
To maximise portability the version of C defined in ISO/IEC 9899:1990
should be used. This is more commonly referred to as C90. 
```

> This question comes up a little now and then, it should probably be part of our FAQ...

I agree, and I suggest putting the above statement also in the `CONTRIBUTING` file.

Yet as mentioned I found today that the current OpenSSL code base does **not** adhere to this at various places - for instance I get:
```
crypto/conf/conf_def.c: In function 'get_next_file':
crypto/conf/conf_def.c:733:29: error: implicit declaration of function 'strcasecmp'; did you mean 'strncmp'? [-Werror=implicit-function-declaration]
         if ((namelen > 5 && strcasecmp(filename + namelen - 5, ".conf") == 0)
                             ^~~~~~~~~~
                             strncmp
```
and
```
In file included from apps/include/apps.h:31,
                 from apps/lib/apps.c:43:
/usr/include/signal.h:321:31: error: unknown type name 'stack_t'
 extern int sigaltstack (const stack_t *__restrict __ss,
                               ^~~~~~~
/usr/include/signal.h:322:4: error: unknown type name 'stack_t'; did you mean 'clock_t'?
    stack_t *__restrict __oss) __THROW;
    ^~~~~~~
    clock_t
```

**Comment by levitte** [comment]
> Well, I find it at odd (by GCC at least) that the C version and type of the compiler used can be detected during compilation while `-pedantic` apparently cannot, though all of these have an effect on the available features.

You're missing the point that it's not `-pedantic` that control what features are available, but the C language standard you apply, i.e. `-std=c90`, `-ansi`...  `-pedantic` adds warnings *corresponding to that language standard*.

As for `-pedantic` not generating a macro, I don't find that odd at all.  `-pedantic` is a warning flag in the same vein as `-Wxxx`...  as a matter of fact, `-pedantic` is a synonym for `-Wpedantic`.  You don't see the compiler generate all sorts of flag macros for all sorts of warnings, do you?  If the compiler would add macros for all the warnings, it would also make the warnings pointless, as they would be too simple to avoid with a macro definition check.

Simply put, if you don't want pedantic warnings, don't use `-pedantic`.  If you want `OPENSSL_FUNC` to be guaranteed to give an actual function name, set your compiled to an appropriate language standard (at least c94, obviously)...  and you still have the source file name and line number, unless you turned that off.

**Comment by levitte** [comment]
As for adding a comment, I'll gladly do that

**Comment by levitte** [comment]
> As for adding a comment, I'll gladly do that

Done.

I also removed the use of `__FUNCSIG__`, as it turns out to be a "pretty" variant of `__FUNCTION__` with MSVC, and only makes a difference with C++.

**Comment by DDvO** [comment]
Well, correct that officially speaking it's not `-pedantic` but the C standard+version that controls the availability, although in fact (with GCC at least) it does control the sloppiness that allows to use `__FUNCTION__` also for GNU/C90.

I tend to agree on your comment regarding the detectability of various compiler warning options.
Yet GCC's `-pedantic` unfortunately is not _entirely_ like `-Wpedantic` - I've experimented with 
```
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
#define OPENSSL_FUNC __FUNCTION__
#pragma GCC diagnostic pop
```
since `#pragma GCC diagnostic` is supposed to work for any compiler warning types, 
but for `-Wpedantic` this leads to a compiler error and still does not allow the use of `__FUNCTION__` with the '90 version of C.

File: include/openssl/macros.h [hunk 1]
```diff
@@ -131,6 +131,22 @@
 #  endif
 # endif
 
+/*
+ * __func__ was standardized in C99, so for any compiler that claims
+ * to implement that language level or newer, we assume we can safely
+ * used that symbol.
```
Comment (DDvO) [inline include/openssl/macros.h]: typo: used -> use

File: include/openssl/macros.h [hunk 2]
```diff
@@ -131,6 +131,22 @@
...
+ *
+ * GNU C also provides __FUNCTION__ since version 2, which predates
+ * C99.  We can, however, only use this if __STDC_VERSION exists,
```
Comment (DDvO) [inline include/openssl/macros.h]: typo: `__STDC_VERSION` -> `__STDC_VERSION__`

**Comment by DDvO** [comment]
Thanks for adding that comment (while this does not directly give the hint that where needed, one can override OPENSSL_FUNC).

Could you also add to `CONTRIBUTING` that any new code must adhere to the ISO C90 standard
and that this can be checked with GCC and `-std=gnu90` (or as would be even better, `-std-=c90`) ?

**Comment by levitte** [comment]
> I've experimented with 
> ```
> #pragma GCC diagnostic push
> #pragma GCC diagnostic ignored "-Wpedantic"
> #define OPENSSL_FUNC __FUNCTION__
> #pragma GCC diagnostic pop
> ```

I assume that's not live code.  The warning isn't given by the preprocessor (i.e. when `OPENSSL_FUNC` is defined) but by the compiler (i.e. when `OPENSSL_FUNC` is *used*).

**Comment by richsalz** [comment]
Perhaps a travis build to explicitly check std90 or whatever?

**Comment by levitte** [comment]
in bar.c:

``` C
#include <stdio.h>

#define OPENSSL_FUNC __FUNCTION__

int main() {
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
  printf("OPENSSL_FUNC = %s\n", OPENSSL_FUNC);
#pragma GCC diagnostic pop
}
```

```
: ; gcc -ansi -pedantic -o bar bar.c
: ; ./bar
OPENSSL_FUNC = main
: ; gcc -ansi -Wpedantic -o bar bar.c
: ; ./bar
OPENSSL_FUNC = main
```

**Comment by levitte** [comment]
> Perhaps a travis build to explicitly check std90 or whatever?

We do have one checking `-ansi`, which is the same thing.  There are some extra configuration options in there, which might possibly circumvent certain warnings (not sure how wise *that* is)

File: include/openssl/macros.h (hunk 1 above)
Comment (levitte) [reply include/openssl/macros.h]: Fixed

File: include/openssl/macros.h (hunk 2 above)
Comment (levitte) [reply include/openssl/macros.h]: Fixed

**Comment by levitte** [comment]
> Could you also add to `CONTRIBUTING` ...

I would rather see that we do that in another PR.

**Comment by DDvO** [comment]
> > I've experimented with
> > ```
> > #pragma GCC diagnostic push
> > #pragma GCC diagnostic ignored "-Wpedantic"
> > #define OPENSSL_FUNC __FUNCTION__
> > #pragma GCC diagnostic pop
> > ```
> 
> I assume that's not live code. The warning isn't given by the preprocessor (i.e. when `OPENSSL_FUNC` is defined) but by the compiler (i.e. when `OPENSSL_FUNC` is _used_).

Ah, right, when using the pragmas where the OPENSSL_FUNC macro is used (not where it is defined), it does work. Thanks also for your `bar.c` example, with which I've been able to reproduce that.

So unfortunately those pragmas do not really help since it would have to be applied for all usages of the macro.

**Comment by levitte** [comment]
> So unfortunately those pragmas do not really help since it would have to be applied for all usages of the macro.

Yes...  and the answer goes back to not using something the language standard can't promise.

**Comment by levitte** [comment]
Ping, @openssl?  We're pretty much done here, methinks.

**Comment by DDvO** [comment]
> > Perhaps a travis build to explicitly check std90 or whatever?
> 
> We do have one checking `-ansi`, which is the same thing. There are some extra configuration options in there, which might possibly circumvent certain warnings (not sure how wise _that_ is)

I'd suggest having at least one Travis build where just `std=c90` is used, to catch the compatibility issues I've (partly) quoted above.

**Comment by petrovr** [comment]
David von Oheimb wrote:
> As I wrote in #9911, also this version won't work when configuring with `-std=gnu90` (or `-std=c90`) and `-pedantic`.
No it will pass .
It is easy to test with this project https://gitlab.com/e_nss .

What will fail is '-std=gnu90 -pedantic -Werror'.

In config.log we could see :
...
conftest.c:28:15: error: ISO C90 does not support '__func__' predefined 
identifier [-Werror=pedantic]
...
conftest.c:28:15: error: ISO C does not support '__FUNCTION__' 
predefined identifier [-Werror=pedantic]
...
So after configure config.h has following:
...
#define __func__ ""
...

So far so good but later build fail
...
.../openssl/build/master_inplace/include/openssl/sha.h:91:36: error: ISO 
C90 does not support 'long long' [-Werror=long-long]
  #  define SHA_LONG64 unsigned long long
                                     ^
...


With other words OpenSSL requires some post C89 functionality and 
discussion for support of either __func__ or __FUNCTION__ makes no sense.



For protocol :
$ gcc --version
gcc (GCC) 5.5.0


Regards,
Roumen Petrov

**Comment by DDvO** [comment]
@petrovr, yes, but 

* we already clarified that the fix in this PR makes the build pass regardless of the C version and independent of `-pedantic` (BTW, even in the presence of `-Werror`), 
though at the expense of making `OPENSSL_FUNCTION` useless for C version 89/90 (unless specific remedy is added by the user).

* I also already noted that various other parts of current OpenSSL does not really comply with C90 (yet this is a more general topic, likely better to be followed up elsewhere).

**Comment by DDvO** [comment]
For the more general C90 compliance issue and more explicit documentation of C version requirements I've just set up an independent issue #9928.

**Comment by DDvO** [comment]
How to proceed with this PR?
Can I (as a normal contributor) formally approve it?

Since the situation reported in #9911 is blocking us from using `std=gnu90' together with --strict-warnings` I'm ok with the strict interpretation of not relying on the unofficial GNU support of `__FUNCTION__` in pre-99 C despite its pragmatic deficiency discussed above.

**Comment by levitte** [comment]
Not formally, no. You'd have to be an @openssl/committers member at least.
Speaking of, I would like to see a reaction. This is fairly low hanging fruit!

File: include/openssl/macros.h [hunk 3]
```diff
@@ -131,15 +131,31 @@
...
-#  elif defined(__STRICT_ANSI__)
-#   define OPENSSL_FUNC "(unknown function)"
-#  elif defined(_MSC_VER) || (defined(__GNUC__) && __GNUC__ >= 2)
-#   define OPENSSL_FUNC __FUNCTION__
-#  elif defined(__FUNCSIG__)
-#   define OPENSSL_FUNC __FUNCSIG__
+#  if defined(__STDC_VERSION__)
+#   if __STDC_VERSION__ >= 199901L
+#    define OPENSSL_FUNC __func__
+#   elif defined(__GNUC__) && __GNUC__ >= 2
```
Comment (paulidale) [inline include/openssl/macros.h]: I need to read the standard to see why this doesn't produce a syntax error when `__GNUC__` isn't defined.....

**Review by paulidale — APPROVED** [review]

File: include/openssl/macros.h (hunk 3 above)
Comment (bernd-edlinger) [reply include/openssl/macros.h]: A macro that is not defined is evaluates to 0.
There is only an issue of the undefined macro is used with parentheses,
since 0(x) is bad and gives an error

Comment (paulidale) [reply include/openssl/macros.h]: Thanks for the explanation.  I wasn't looking forward to digging into the standard again.

**Comment by levitte** [comment]
Merged.

ec87a649dd2128bde780f6e34a4833d9469f6b4d include/openssl/macros.h: Rework OPENSSL_FUNC for div C standards

**Comment by levitte** [comment]
Thanks!

**Comment by mspncp** [comment]
> Can I (as a normal contributor) formally approve it?

That depends on what you mean with 'formally': Technically, GitHub will allow any user  to do reviews and approve or request changes. They will appear among the others in the top right reviewer list, only greyed out. We do appreciate reviews by contributors and take them into account, even more if we know they have expertise on the subject. Only formally, the approvals don't count for the merge precondition (two committer approvals, including one OMC approval), as @levitte already said.

**Comment by richsalz** [comment]
Out  of curiosity (I know this is a closed issue) what does 'div' mean in "for div C standards" ?

**Comment by mspncp** [comment]
I would say, it means 'diverse'.

**Comment by DDvO** [comment]
Thanks @mspncp for your additional info regarding 'non-counting' approvals.
I'm going to try this out next time when I come across a PR that I'd like to see merged after having done an 'informal' review.

**Comment by slontis** [comment]
Somehow this has broken all solaris builds...

cc  -I. -Iinclude -Iapps/include -I.. -I../include -I../apps/include  -KPIC -xarch=generic64 -xstrconst -Xa -xO5 -xdepend -xbuiltin -m64 -DFILIO_H -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/export/home/nse/workspace/openssl-solaris11_x64_sunc515/install/ssl\"" -DENGINESDIR="\"/export/home/nse/workspace/openssl-solaris11_x64_sunc515/install/lib/engines-3\"" -DMODULESDIR="\"/export/home/nse/workspace/openssl-solaris11_x64_sunc515/install/lib/ossl-modules\"" -D_REENTRANT -DNDEBUG   -c -o apps/lib/libapps-lib-apps.o ../apps/lib/apps.c
"../apps/lib/apps.c", line 1209: warning: assignment type mismatch:
	pointer to pointer to const char "=" pointer to const void
"../apps/lib/apps.c", line 1210: warning: assignment type mismatch:
	pointer to pointer to const char "=" pointer to const void
"../apps/lib/apps.c", line 1210: warning: assignment type mismatch:
	pointer to pointer to const char "=" pointer to const void
"../apps/lib/apps.c", line 1211: warning: assignment type mismatch:
	pointer to pointer to const char "=" pointer to const void
"../apps/lib/apps.c", line 1212: warning: assignment type mismatch:
	pointer to pointer to const char "=" pointer to const void
"../apps/lib/apps.c", line 1212: warning: assignment type mismatch:
	pointer to pointer to const char "=" pointer to const void
"../apps/lib/apps.c", line 1398: undefined symbol: OPENSSL_FUNC
"../apps/lib/apps.c", line 1398: warning: improper pointer/integer combination: arg #3
cc: acomp failed for ../apps/lib/apps.c

**Comment by slontis** [comment]
I assume this means that one of __FUNCTION__ , __func__, etc is not defined on solaris now with this new logic.

Should it be doing this kind of pattern...
#if defined(__func__)
#define OSSL_FUNC __func__
......

It at least needs
#ifndef OPENSSL_FUNC
#define OPENSSL_FUNC "unknown"
#endif

**Comment by levitte** [comment]
Fix in #9976

**Comment by levitte** [comment]
Yes, exactly
